src/error.rs

```diff
@@ -6,73 +6,108 @@ use std::error::Error as StdError;
 #[cfg(feature = "std")]
 use self::super::TriviallyTransmutable;
 use core::fmt;
+#[cfg(feature = "std")]
+use core::marker::PhantomData;
 
 
 /// A transmutation error. This type describes possible errors originating
-/// from operations in this crate.
+/// from operations in this crate. The two type parameters, when defined,
+/// represent the source element type and the target element type
+/// respectively.
 ///
 /// # Examples
 ///
 /// ```
 /// # use safe_transmute::{ErrorReason, Error, transmute_bool_pedantic};
-/// # unsafe {
 /// assert_eq!(transmute_bool_pedantic(&[0x05]), Err(Error::InvalidValue));
-/// # }
 /// ```
-#[derive(Clone, Debug, PartialEq, Eq, Hash)]
-pub enum Error {
+#[derive(Clone, PartialEq, Eq, Hash)]
+pub enum Error<T = (), U = ()> {
     /// The data does not respect the target type's boundaries.
     Guard(GuardError),
     /// The given data slice is not properly aligned for the target type.
     Unaligned(UnalignedError),
-    /// The given data vector is not properly aligned for the target type.
+    /// The given data vector is not properly aligned in memory for the target type.
     ///
     /// Does not exist in `no_std`.
     #[cfg(feature = "std")]
     UnalignedVec(UnalignedVecError),
+    /// The data vector's element type does not have the same size and minimum
+    /// alignment as the target type.
+    ///
+    /// Does not exist in `no_std`.
+    #[cfg(feature = "std")]
+    IncompatibleVecTarget(IncompatibleVecTargetError<T, U>),
     /// The data contains an invalid value for the target type.
     InvalidValue,
+
+    #[cfg(not(feature = "std"))]
+    #[doc(hidden)]
+    None(::core::marker::PhantomData<(T, U)>),
+}
+
+impl<T, U> fmt::Debug for Error<T, U> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            Error::Guard(e) => write!(f, "Guard({:?})", e),
+            Error::Unaligned(e) => write!(f, "Unaligned({:?})", e),
+            Error::InvalidValue => f.write_str("InvalidValue"),
+            #[cfg(feature = "std")]
+            Error::UnalignedVec(e) => write!(f, "UnalignedVec({:?})", e),
+            #[cfg(feature = "std")]
+            Error::IncompatibleVecTarget(_) => f.write_str("IncompatibleVecTarget"),
+            #[cfg(not(feature = "std"))]
+            Error::None(_) => unreachable!(),
```

Comment:
I'd rather these not panic, tbh, or at least not with an unreachable

```rust
#[test]
fn damn_danielle() {
    use core::fmt::{Write, Error as FError};
    struct A;
    impl Write for A {
        fn write_str(&mut self, s: &str) -> Result<(), FError> {
            Ok(())
        }
    }
    write!(&mut A, "{:?}", Error::None(::core::marker::PhantomData::<((), ())>));
}
```

Reply:
It is not meant to be used, but you're right. I wouldn't want this to trigger a panic inside another panic.